### ISSUE TITLE
Incorrect determination of the duration of the tracks in some CUE

### ISSUE THREAD

**Issue by komarevtseff** [cmus/cmus#251]
For example CUE:
https://bpaste.net/show/89d2ed623967
In the first three tracks duration faithful, and then there is no.

For clarity screenshots for this album in DeaDBeeF (true) and cmus (false):
![2015-02-16-062548_1366x751_scrot](https://cloud.githubusercontent.com/assets/3647789/6206903/8c52db0e-b5a5-11e4-8d66-4e71f96a9f87.png)
![2015-02-16-062532_1088x456_scrot](https://cloud.githubusercontent.com/assets/3647789/6206919/157b2daa-b5a6-11e4-9216-18e153244252.png)

**Comment by flyingmutant** [comment]
Thanks for reporting! We use `libcue` for CUE parsing — and it looks like this is another little problem with it.

**Comment by flyingmutant** [comment]
Pasting the CUE in question here for reference:

```
REM ACCURATERIPID 0016fab2-00f67491-c30ac90e
REM GENRE "Alternative Rock/Nu Metal/Rap Rock"
REM DATE 2000
PERFORMER "Linkin Park"
TITLE "Hybrid Theory"
FILE "Linkin Park - Hybrid Theory.wv" WAVE
  TRACK 01 AUDIO
    TITLE "Papercut"
    PERFORMER "Linkin Park"
    INDEX 01 00:00:00
  TRACK 02 AUDIO
    TITLE "One Step Closer"
    PERFORMER "Linkin Park"
    INDEX 01 03:05:40
  TRACK 03 AUDIO
    TITLE "With You"
    PERFORMER "Linkin Park"
    INDEX 01 05:41:17
  TRACK 04 AUDIO
    TITLE "Points of Authority"
    PERFORMER "Linkin Park"
    INDEX 01 09:04:35
  TRACK 05 AUDIO
    TITLE "Crawling"
    PERFORMER "Linkin Park"
    INDEX 00 12:15:59
    INDEX 01 12:24:55
  TRACK 06 AUDIO
    TITLE "Runaway"
    PERFORMER "Linkin Park"
    INDEX 00 15:46:41
    INDEX 01 15:53:55
  TRACK 07 AUDIO
    TITLE "By Myself"
    PERFORMER "Linkin Park"
    INDEX 00 18:57:49
    INDEX 01 18:57:50
  TRACK 08 AUDIO
    TITLE "In the End"
    PERFORMER "Linkin Park"
    INDEX 00 20:34:38
    INDEX 01 22:07:35
  TRACK 09 AUDIO
    TITLE "A Place for My Head"
    PERFORMER "Linkin Park"
    INDEX 00 24:52:38
    INDEX 01 25:43:57
  TRACK 10 AUDIO
    TITLE "Forgotten"
    PERFORMER "Linkin Park"
    INDEX 00 28:28:43
    INDEX 01 28:48:27
  TRACK 11 AUDIO
    TITLE "Cure for the Itch"
    PERFORMER "Linkin Park"
    INDEX 00 32:02:56
    INDEX 01 32:02:62
  TRACK 12 AUDIO
    TITLE "Pushing Me Away"
    PERFORMER "Linkin Park"
    INDEX 00 34:04:04
    INDEX 01 34:40:00
  TRACK 13 AUDIO
    TITLE "My December [bonus track]"
    PERFORMER "Linkin Park"
    INDEX 00 37:51:33
    INDEX 01 37:53:20
  TRACK 14 AUDIO
    TITLE "High Voltage [bonus track]"
    PERFORMER "Linkin Park"
    INDEX 00 42:13:19
    INDEX 01 42:15:40
```

**Comment by Smattr** [comment]
I think the underlying issue here is that libcue does not treat index 0 as the pregap. Reading the cue spec  leads me to think `INDEX 00` is intended to be an alias for `PREGAP`. However, libcue seems to treat it identically to any other index. To quote the relevant libcue 1.4.0 code at time of writing (src/libcue/cue_parser.y:235-254):

```
    | INDEX NUMBER time '\n' {
        int i = track_get_nindex(track);
        long prev_length;

        if (0 == i) {
            /* first index */
            track_set_start(track, $3);

            if (NULL != prev_track && NULL == cur_filename) {
                /* track shares file with previous track */
                prev_length = $3 - track_get_start(prev_track);
                track_set_length(prev_track, prev_length);
            }
        }

        for (; i <= $2; i++)
            track_add_index(track, \
            track_get_zero_pre(track) + $3 \
            - track_get_start(track));
    }
```

IIUC, this code should be special casing when `$2 == 0` and adjusting the track's pregap instead of adding a new index.

I realise I'm coming in part way through this issue and the above is based on my own partial understanding of libcue and the OP's desired behaviour, so apologies for time wasting if my comments are incorrect.

**Comment by flyingmutant** [comment]
@Smattr, thanks for the insight — you are probably right about this. Unfortunately, not much we can do expect drop the libcue dependency and rewrite CUE parsing from scratch.

**Comment by Smattr** [comment]
I was actually thinking of fixing this in libcue, though it's not clear to me whether libcue has an active maintainer. In the meantime @lipnitsk [posted](http://sourceforge.net/p/libcue/discussion/934048/thread/a3e9844b/) that they've started trying to make these changes over [here](https://github.com/lipnitsk/libcue).

Interestingly [cuetools](https://github.com/svend/cuetools), the project libcue was extracted from, has identical code here and its utility cuebreakpoints reports the correct sequence points for the above input. The approach taken is to just ignore the index 0 entirely and just look at index 1. At this point, it's ambiguous to me what the "correct" behaviour is. If you want to change to the cuetools approach, let me know and I'll see if I can put together a PR that does this. I don't know the internals of cmus particularly well, but it would be a good chance to learn.

**Comment by flyingmutant** [comment]
Fixing it in upstream libcue is great (but can be difficult is the project is abandoned) — right now cmus uses `track_get_start()` function, which, as we can see, has a bug. However, I think going with cuetools approach is a fine course of action for cmus. PR will be greatly appreciated (you should probably start with looking at `cue.c`).

**Comment by lipnitsk** [comment]
I believe @jotrk is the original maintainer of libcue. I have sent him a request to move over libcue development to github, but have not heard a response yet. So I went ahead and forked libcue for now. I plan to clean it up and eventually release version 2.0.0 with minor API/ABI changes.

That said, is the current version of cuetools a superset of what libcue provides and does it fix the known libcue issues?

**Comment by Smattr** [comment]
From what I understand, libcue was originally extracted from cuetools. Cuetools contains more than just libcue and it looks to me like it has had extra commits that were never merged into libcue. If libcue were recreated from the current cuetools, we would still have the same issue, as the code in this area is identical.

@lipnitsk, maybe check with @svend, the maintainer of cuetools, as to their feelings about the future of libcue and its relationship to cuetools going forwards. @flyingmutant, I'll hold off on any cmus PR for now until we get some clarity on the future of libcue.

**Comment by jotrk** [comment]
Quick tour about the history and relationship of libcue and cuetools.

Back in 2006 I wanted to have support for CUE sheets in MPD. There was no library available and the only program which actually was capable of handling CUE sheets where cuetools.

Since I didn't want to just copy and paste the whole code in MPD, I decided to create a library. At that time cuetools looked pretty much abandoned, so I wrote a mail to @svend, but never got a reply. I decided to relicense libcue under BSD, because I rewrote it in large parts.

**Comment by Wang-Yue** [comment]
I can confirm @lipnitsk fixed the problem. 

cmus needs to add a pkg-config check to detect libcue version, as libcue2's header is <libcue.h> rather than <libcue/libcue.h>

**Comment by Wang-Yue** [comment]
See  PR #384

**Comment by flyingmutant** [comment]
Fixed by #384 (libcue2).
